### DISCUSSION TITLE
리마인더 기능에 대한 토론

### DISCUSSION THREAD

**Discussion by Choi-JJunho** [woowacourse-teams/2023-pium#143] [뜨거운 토론]
## 논제

리마인더의 다음 기능들에 대한 구체적인 방향성을 잡고 간다.

- 미루기
- 물주기
- 날짜 지정 물주기

<img width="423" alt="image" src="https://github.com/woowacourse-teams/2023-pium/assets/49794401/fa7b3a17-d496-47db-af9f-ce6e9d8ec837">

### 미루기 버튼은 어떻게 동작해야하는가?

- 언제?
  - ex) `지각` / `오늘 할일` / `미래에 해야할 일` 모두 미루기 버튼이 존재한다.
- 어떻게?
  - ex) 미루기를 누르면 `nextWaterDate`가 `Today + 1일`이 된다.
  - ex) 미루기를 누르면 `달력에서 선택한 날짜 + 1일`이 된다.

### 물주기 버튼은 어떻게 동작해야하는가?

`지각` / `오늘 할일` / `미래에 해야할 일` 3가지 상황에서 날짜 지정 물주기는 언제 어떤 방식으로 제공해야하는가?

- 언제?
  - ex) 지각, 오늘 할일에 대해서만
- 어떤 방식으로?
  - ex) lastWaterDate와 Today 사이에 대해서만 날짜를 활성화하여 선택할 수 있도록 한다.

**Comment by WaiNaat** [comment]
미루기라는 행위의 주 목적에 따라 다를거같아요
흙이 안 말랐을 때 내일 다시 사용자가 확인했으면 하는 경우에는 백엔드에서 그냥 nextwaterdate+하루
출장 등으로 며칠씩 자리를 비워서 못준다면 프론트가 날짜를 계산해서 보내기

저는 개인적으로는 전자가 맞다고 생각합니다

**Comment by yeonkkk** [comment]
물주기 기능 통합에 대한 부분도 회의에서 언급되어 같이 작성합니다.

### 물주기 기능 통합
1. 미루기 버튼과 날짜 선택 버튼이 붙어있어서, 의도와 다르게 미룰 날짜를 선택할 수 있는 것 처럼 보일 것 같아요.
2. 이와 반대로 오늘 물주기 버튼과 날짜 선택 버튼의 위치가 너무 떨어져있고, 오늘 물주기 버튼이 직관적이지 않은 것 같아요.

→ 차라리 `오늘 물주기`와 `날짜 선택`을  `물주기`로 합치는게 좋을 것 같습니다. 물주기 버튼을 누르면 날짜 선택을 해서 기능이 동작하면 좋을 것 같아요.  


### 미루기
1. 기존에서 서버는 미루기 로직을 `NextWaterDate + 1` 로 처리했습니다.
2. 만약 미루기 기능을 `지각`에도 도입할거라면 `프론트엔드에서 보내주는 날짜 +1`로 처리하면 좋을 것 같네요.  그런데 미루기 기능을 `오늘 할 일`과 `미래 할 일`에만 한정에서 적용한다면 `NextWaterDate + 1` 로 처리하는게 좋을 것 같아요.

### 물주기 버튼 동작
- 아직 물주기 날짜가 남은 식물들도 상태에 따라 미리 물주는 상황을 고려해보면 지각, 오늘 할 일, 미래 할 일 모든 곳에서 버튼이 동작해야한다고 생각합니다. 
- 마지막으로 물 준 날짜와 오늘 사이에 물 주는게 가능하도록 해야할 것 같습니다.

**Comment by hozzijeong** [comment]
+) 추가로 날짜 선택으로 물을 주는 경우에 유효한 물주기 날짜 범위를 제공해야 하는데, 이 기능을 완벽하게 지정하려면 데이터에서 필드 한개를 더 받아와야 합니다. 이전 날짜라고 생각해서 prevWaterDate라고 하겠습니다.

지각인 경우: 원래 물을 주기로 했던 날(nextWaterDate) ~  today에만 물을 줄 수 있도록 설정
오늘인 경우: 이전에 물을 줬던 날 (prevWaterDate) ~ today에만 물을 줄 수 있도록 설정
앞으로 물을 줘야하는 경우: prevWaterDate ~ today에만 물을 줄 수 있도록 설정

즉, 이전에 물을 줬던 날짜도 필드에 저장되어야 할 것 같아요

**Reply by Choi-JJunho** [reply]
prevWaterDate == lastWaterDate로 이해해도 될까요?

# 반려 식물 단건 조회 데이터

```json
{
  "id": 2,
  "nickname": "기영이",
  "imageUrl": "https://image.com",
  "dictionaryPlant": {
    "id" : 1,
    "name" : "스투키"
  },
  "location": "거실",
  "flowerpot": "플라스틱",
  "light": "창 밖",
  "wind": "바람이 안통해요",
  "birthDate": "2023-07-08",
  "lastWaterDate": "2023-07-10", // 마지막으로 물을 준 날 == 이전에 물 줬던 날
  "waterCycle": 7,
  "dDay": 3, // nextWaterDay를 dDay로 수정
  "daySince": 95,
  "nextWaterDate": "2023-07-10" // 추가됨
}
```

**Reply by hozzijeong** [reply]
아하 넵넵 맞습니다! 리마인더 구현을 할 때 lastwaterdate가 추가로 들어갈 것 같아서 말씀 드린것이었습니다.

**Reply by yeonkkk** [reply]
@hozzijeong 
클린, 리마인더 조회 응답에 아래처럼 추가해달라는 말 맞나요?  
맞으면 api 명세 수정하고 저도 적용하겠습니다.

```json
{
	"data": [
		{
			"petPlantId": 1,
			"image": "image.com",
			"nickName": "참새나무",
			"dictionaryPlantName": "알로카시아",
			"dDay": 3, //  음수면 할일 | 0이면 오늘 | 양수면 지각
			"nextWaterDate": "2023-07-07",
                         "lastWaterDate": "2023-07-10" // 추가
		},
		{
			"petPlantId": 2,
			"image": "image.com",
			"nickName": "그레이나무",
			"dictionaryPlantName": "스투키",
			"dDay": 3, //  음수면 할일 | 0이면 오늘 | 양수면 지각
			"nextWaterDate": "2023-07-16",
                         "lastWaterDate": "2023-07-10" // 추가
		}
	]
}

```

**Reply by hozzijeong** [reply]
맞습니다! 늦게확인해서 죄송합니다 ㅠ

**Comment by bassyu** [comment]
물주기 통합 괜찮은 것 같아요!
미루기도 기능을 단순하게 가져가면 좋을 것 같아요!

### 물주기 버튼

(디폴트) 오늘
(옵션) 과거선택 가능

디자인은 대충..

( `오늘` 물주기 ) <- 여기서 오늘을 클릭하면 과거 날짜 선택 가능

### 미루기 버튼

오늘 / 지각 에서만 존재하는 버튼

지각이든 오늘이든 `오늘 + 1`로 미루기

**Reply by Choi-JJunho** [reply]
오 좋은데요?
저는 쵸파와 같은 의견입니다~!

**Comment by WaiNaat** [comment]
### 미루기

오늘/지각일 때 `오늘 + 1`로 미루는게 좋은거같아요
미래는 `nextWaterDate + 1`로 미루는게 좋은거같아요
만약 경우를 나누기 힘들다면 `nextWaterDate + 1`이 편할거같아요

### 물주기

저는 날짜 선택이랑 물주기 버튼은 나눠져 있어야 한다고 생각해요
합쳐버리면 오늘 물을 주는 상황에서도 날짜 선택 -> 완료라는 절차가 필요한데 리마인더에서 완료를 위해 두 단계를 거치는 건 너무 많아 보입니다

**Reply by WaiNaat** [reply]
![image](https://github.com/woowacourse-teams/2023-pium/assets/77872742/4c91bb7f-9ab8-4451-8dfa-4d3df4c9ba2a)

대충 이런느낌으로 통합해도 괜찮을거같아요 디자인은 옛날버전이라 기능만 봐주세요..
날짜는 기본값은 '오늘'이고 클릭해서 `nextWaterDate` 부터 '내일'까지 선택할 수 있구요
'물주기' 버튼으로 서버에 물주기 요청을 보냅니다. 날짜가 '내일'이라면 '미루기'로 바뀝니다

**Reply by bassyu** [reply]
오 제가 생각했던 버튼 모양이 딱 참새가 보여준 디자인이네요

**Reply by WaiNaat** [reply]
근데 가로로 하면 좀 공간이 많이들어서 세로로 가는게 좋을거같아요

**Reply by bassyu** [reply]
오 그것도 좋은데요?

**Reply by hozzijeong** [reply]
UI가 많이 복잡해지네요.. 로직도 좀 꼬일것 같기도 하고요...

일단은 모든 상태(지각, 오늘, 미래)에 대해서는 물주기와 미루기 기능이 있어야 합니다. (과거의 경우, 당장에 어제라고 해도, 어제 물을 안줘서 오늘 주려고 봤는데 흙이 촉촉한 경우 미루기가 필요)

하지만, 참새가 제안해 준 기능을 보면 과거의 미루기는 불가능해 보이네요...

또한 오늘을 클릭하지 않으면 미루기라는 기능이 있는지도 잘 모를것 같기도 합니다 ㅠ

**Reply by hozzijeong** [reply]
![image](https://github.com/woowacourse-teams/2023-pium/assets/50974359/8fa441ef-a686-464e-9a83-66b35f713b83)
작게나마 만들어보본 기능인데, 위와 같이 되면 많이 이상할까요?

오늘 물주기가 추가되면 좋겠지만, 디자인이 어떻게 되면 좋을지 모르겠네요...

**Comment by hozzijeong** [comment]
조이: 날짜 선택과 미루기가 붙어있어서 미루기가 날짜선택인줄 알았음. 물주기를 나누고, 물주기를 클릭했을 때 날짜 버튼이었으면 좋겠음. 또한 물주기 버튼이 직관적이지 않다고 생각함.

미루긱는 기존에 nextWaterDate +1 이라 생각했음. 또한 지각에 미루기 버튼이 있을거라 생각을 못했음. 도입을 한다면 지각이랑 오늘 할일의 미루기는 오늘 +1 미래는 미래에 날짜 +1 로 바꾸는게 어떤지? 

그레이: 미래의 미루기는 ㅜ머죠? 
ㄴ 미래: 원래 물주기 날짜 +1 / 오늘 + 지각 : 오늘 +1

클린: 오늘 물주기에 대한 생각을 하지 말고, 물주기 선택을 눌렀을 때 날짜 선택하는 기능이 되면 좋겠다. 

쵸파: 한번에 클릭하고 날짜를 선택할 수 있으면 날자가 분리되어야 한다고 생각함. 다시 생각을 해보니까 리마인더를 쓴다고 했을 때 너무 편의성을 강조하지 않아도 될까라는 생각도 했고, 안써봐서 모르겠지만 당일 기록 보다는 일기 처리하듯이 처리하는게 더 많을 듯. 기록을 좀 더 쉽게하고 UI등을 종합적으로 생각해봤을 때 물주기를 클릭했을 때 날짜를 선택하면 더 좋을 것 같긴함. 미루기 역시 달력이 똑같이 나오는 건 어떤지?

참새: 제 의견은 사실 저는 날짜 선택 뎁스 추가가 리마인더에서는 치명적이라 생각함. 그 생각은 같음. 쵸파 생각을 듣고 보니 핵심 기능은 아니라 많ㅁ은 고민 시간을 투자하는게 손해같음. 날짜 선택을 해서 하는 방향도 괜찮을 것 같은데 이제 저 달력 UI에서 날짜를 선택하는게 얼마나 편할지 모르겠음. 식물 등록같은 경우에는 날자를 선택하고 다음을 누르면 보내져서, 여기는 onChange에서 날짜가 넘어가서 실수를 할 가능성이 높음.  -> 토스트로 실행취소를 하면 어떤지? => 실수를 방지할 수 있음. 

기록을 잘못했을 경우에 수정할 수 있는 기능이 필요하다. 

그레이, 주노: 동의

의견을 다 들어봤을 때 물주기, 미루기 달력을 띄우고 validation을 준다가 괜찮아 보임. 

그렇다면 validation을 하기 위해 정보가 좀 더 필요하지 않나? 

### 변경안

미루기 -> 달력이 나타남
물주기 -> 달력이 나타남.

지각, 오늘에 대해서는 미루기. 
미래 -> 날짜 변경.

당일 물주기 버튼이 없어짐. 

이는 모든 변수에 대해 대응이 가능하기 위한 변경 사항임. 


### 검증

#### 물주기

지각일 때 물주기 검증 조건:
min: lastWaterDate 이후
max: 오늘 까지

오늘 물 주는 날:
min: lastWaterDate이후
max: 오늘까지

미래:
min: lastWaterDate 이후
max: 오늘까지

#### 미루기 (날짜 변경)

지각, 오늘 할 일, 미래 :
min: 내일 부터
max: 없음.

**Comment by bassyu** [comment]
# 7월 28일 (금) 회의

조이 : 잘짜 선택과 물주기를 물주기 버튼으로 통일하자
클린 : 조이와 동일합니다. 2뎁스로 가는 것이 큰 손해가 아니라고 느껴져서 손해보다는 득이 더 많다고 생각합니다.

쵸파 : 생각이 바뀜 -> 조이 클린과 동일함
기록이 중요한 서비스이기 때문에 이런 방향성도 좋아보인다.
실제로 오늘 당장 물주는 날이 더 많을 지는 모르겠다.

참새: 조이 클린처럼 리마인더 입장에서는 치명적이지만 서비스의 핵심이 아니라 날짜 선택해도 좋을 것 같습니다.
날짜를 실수로 잘못 선택했을 때가 걱정돼요 -> 실행취소 토스트로 해결될 것 같습니다.

주노, 그레이: 동의합니다.

### 공통

하마드 : 이미 물주기를 기록했는데 알고보니 다른 날짜여서 수정하고 싶다면?
반려식물 상세에서 해결할 수 있도록 해야한다.

조이 : 근데 그때 그때 하루씩 미뤄야 원래 기획의 의미에 맞지 않을까요?
클린 : 물론 편의성에서는 그게 맞지만 조금 내려놓더라도 범용성을 챙길 수 있다.
예를들면 물주려는데 3일은 미뤄야 될 것 같다.

각각 검증 조건

모든 날짜의 물주기 : lastWaterDate + 1 ~ 오늘
과거의 미루기 : 내일 ~
오늘의 미루기 : 내일 ~
미래의 미루기 : 내일?? ~

그런데 땡기기도 필요하지 않나요?
미래의 계획을 앞당겨야 할 수도 있잖아요

점점 캘린더의 성격이 짙어지는 느낌이 들긴 하지만 필요한 기능은 맞는 것 같아요

`지각` `오늘`은 미루기 버튼이고 `미래`는 날짜변경 버튼으로 바꾸는 것 어떤가요?
동의합니다!

### 결론

모든 날짜의 물주기 : lastWaterDate + 1 ~ 오늘
과거의 미루기 : 내일 ~
오늘의 미루기 : 내일 ~
미래의 ~미루기~ 날짜변경 : 내일 ~

**Comment by Choi-JJunho** [comment]
리마인더 기능 개발 완료로 close하겠습니다.
